src/dmd/lexer.d

```diff
@@ -194,32 +196,214 @@ unittest
-        format = format string for error
+        format = format string for supplemental message
+        ... = format string arguments
+    */
+    final void errorSupplemental(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        errorSupplemental(loc, format, args);
+        va_end(args);
+    }
+
+    /// ditto
+    abstract void errorSupplemental(const ref Loc loc, const(char)* format, va_list);
+
+    /**
+    Reports a warning message.
+
+    Params:
+        loc = Location of warning
+        format = format string for warning
+        ... = format string arguments
+    */
+    final void warning(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        warning(loc, format, args);
+        va_end(args);
+    }
+
+    /// ditto
+    abstract void warning(const ref Loc loc, const(char)* format, va_list args);
+
+    /**
+    Reports additional details about a warning message.
+
+    Params:
+        loc = Location of warning
+        format = format string for supplemental message
         ... = format string arguments
     */
-    abstract void error(Loc loc, const(char)* format, ...);
+    final void warningSupplemental(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        warningSupplemental(loc, format, args);
+        va_end(args);
+    }
+
+    /// ditto
+    abstract void warningSupplemental(const ref Loc loc, const(char)* format, va_list);
+
+    /**
+    Reports a deprecation message.
+
+    Params:
+        loc = Location of the deprecation
+        format = format string for the deprecation
+        ... = format string arguments
+    */
+    final void deprecation(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        deprecation(loc, format, args);
+        va_end(args);
+    }
+
+    /// ditto
+    abstract void deprecation(const ref Loc loc, const(char)* format, va_list args);
+
+    /**
+    Reports additional details about a deprecation message.
+
+    Params:
+        loc = Location of deprecation
+        format = format string for supplemental message
+        ... = format string arguments
+    */
+    final void deprecationSupplemental(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        deprecationSupplemental(loc, format, args);
+        va_end(args);
+    }
+
+    /// ditto
+    abstract void deprecationSupplemental(const ref Loc loc, const(char)* format, va_list);
+}
+
+/**
+Diagnostic reporter which prints the diagnostic messages to stderr.
+
+This is usually the default diagnostic reporter.
+*/
+class StderrDiagnosticReporter : DiagnosticReporter
+{
+    private const Diagnostic useDeprecated;
+
+    private bool hasErrors_;
+    private bool hasWarnings_;
+    private bool hasDeprecations_;
```

Comment:
Wouldn't a counter be more generic?

Reply:
Yes, I didn't think of that because the lexer is currently only storing a boolean to indicate if an error has occurred or not.

Reply:
Fixed.